wasm/validation.cpp

```diff
@@ -0,0 +1,525 @@
+
+            v.push_ctrl(Block{}, params, results);
+        } else if (Loop const *loop = std::get_if<Loop>(&inst)) {
+            std::vector<Instruction> loop_code = loop->instructions;
+
+            loop_code.emplace_back(End{});
+
+            code.insert(code.begin() + i + 1, loop_code.begin(), loop_code.end());
+
+            std::vector<ValueType> params;
+            std::vector<ValueType> results;
+
+            if (ValueType const *vt = std::get_if<ValueType>(&loop->type.value)) {
+                results.push_back(*vt);
+            } else if (TypeIdx const *idx = std::get_if<TypeIdx>(&loop->type.value)) {
+                if (!m.type_section.has_value()) {
+                    return tl::unexpected{ValidationError::TypeSectionUndefined};
+                }
+
+                if (*idx >= m.type_section.value().types.size()) {
+                    return tl::unexpected{ValidationError::TypeUndefined};
+                }
+
+                params = m.type_section.value().types.at(*idx).parameters;
+                results = m.type_section.value().types.at(*idx).results;
+            }
+
+            // The case of an empty block type is handled implicitly by leaving the vectors empty
+
+            v.push_ctrl(Loop{}, params, results);
+        } else if (std::holds_alternative<End>(inst)) {
+            tl::expected<ControlFrame, ValidationError> maybe_frame = v.pop_ctrl();
+
+            if (!maybe_frame.has_value()) {
+                return tl::unexpected{maybe_frame.error()};
+            }
+
+            v.push_vals(maybe_frame.value().results);
+        } else if (Branch const *branch = std::get_if<Branch>(&inst)) {
+            if (v.control_stack.size() < branch->label_idx) {
+                return tl::unexpected{ValidationError::LabelInvalid};
+            }
+
+            tl::expected maybe_vals =
+                    v.pop_vals(v.label_types(v.control_stack.at(v.control_stack.size() - (branch->label_idx + 1))));
+
+            if (!maybe_vals.has_value()) {
+                return tl::unexpected{maybe_vals.error()};
+            }
+
+            v.unreachable();
+        } else if (BranchIf const *branch_if = std::get_if<BranchIf>(&inst)) {
+            if (v.control_stack.size() < branch_if->label_idx) {
+                return tl::unexpected{ValidationError::LabelInvalid};
+            }
+
+            v.pop_val_expect(ValueType::Int32);
+
+            tl::expected maybe_vals =
+                    v.pop_vals(v.label_types(v.control_stack.at(v.control_stack.size() - (branch_if->label_idx + 1))));
+
+            if (!maybe_vals.has_value()) {
+                return tl::unexpected{maybe_vals.error()};
+            }
+
+            v.push_vals(v.label_types(v.control_stack.at(branch_if->label_idx)));
+        } else if (std::holds_alternative<Return>(inst)) {
+            tl::expected maybe_vals = v.pop_vals(v.label_types(v.control_stack.at(0)));
```

Comment:
Is control_stack guaranteed to have at least one value on it here? If so, maybe add an assert to make that clearer?

Reply:
the control stack is guaranteed to always have 1 frame on it, until you're done validating all instructions. but i make that assumption in other places aside from just this one, and there might be more of them when we add the rest of the instructions, so i'll add an assert at the beginning of the loop.